### ISSUE TITLE
Interpolate AUC in COCOMaP

### ISSUE THREAD

**Issue by LukeWood** [keras-team/keras-cv#149]
the original implementation actually does NOT do this, but it is more accurate with respect to the mathematical definition of MaP.

**Comment by LukeWood** [comment]
I'd argue that this is better because we already diverge numerically from cocoeval.py.  No point in not sticking closer to the mathematical definition.

**Comment by LukeWood** [comment]
When we do this, we can add a parameter to determine the interpolation style.  This will allow us to preserve the numerical_tests/ directory.  Let's make the curve interpolation the default though, as it is the more mathematically correct approach.

AUC has an implementation of this:
https://github.com/keras-team/keras/blob/v2.8.0/keras/metrics.py#L2046-L2533

**Comment by LukeWood** [comment]
This would be a great enhancement, but is significantly complex.  an experienced TF user could take this on!

**Comment by bhack** [comment]
> This will allow us to preserve the numerical_tests/ directory.

I think this is important:

1.  We still need to fix bugs that we have in there current implementation
2. We need this for reproducibility. I suppose that performance tables in papers are related to the `cocoeval.py` version.

**Comment by LukeWood** [comment]
> > This will allow us to preserve the numerical_tests/ directory.
> 
> I think this is important:
> 
> 1. We still need to fix bugs that we have in the current implementation here
> 2. We need this for reproducibility. I suppose that performance tables in papers are related to the `cocoeval.py` version.

they are.  But their numbers can't truly be considered ground truth; if I understand correctly, they use the segmentation map to filter area for bounding boxes in ground truths.  This is clearly not correct.

**Comment by LukeWood** [comment]
This is a separate thread of work as this will not impact `numerical_tests/` at all.  If someone wants to pick this up, feel free!

**Comment by bhack** [comment]
> they are. But their numbers can't truly be considered ground truth; if I understand correctly, they use the segmentation map to filter area for bounding boxes in ground truths. This is clearly not correct.

When I reimplement something I need to reproduce paper numbers. When we will have fixed the bugs we could reason about if we want to have a `cocoeval` compatibile flag or not if the drift Is still relevant.

> This is a separate thread of work as this will not impact numerical_tests/ at all.

My comment was strictly related to:

> This will allow us to preserve the numerical_tests/ directory.

**Comment by LukeWood** [comment]
> > they are. But their numbers can't truly be considered ground truth; if I understand correctly, they use the segmentation map to filter area for bounding boxes in ground truths. This is clearly not correct.
> 
> When I reimplement something I need to reproduce paper numbers. When we will have fixed the bugs we could reason about if we want to have a `cocoeval` compatibile flag or not if the drift Is still relevant.
> 
> > This is a separate thread of work as this will not impact numerical_tests/ at all.
> 
> My comment was strictly related to:
> 
> > This will allow us to preserve the numerical_tests/ directory.

My comment is that this should be an optional parameter.  This is distinct from the issue of accuracy of the original implementation.  Let's try to would like to keep discussions focused and issue threads related to the issue. 

If we don't contributors will be confused by the massive amount of information.  A contributor who wants to tackle this issue needs to know that this should be implemented behind a constructor parameter controlling controlling the interpolation mode; not whether or not the numerical tests are highly accurate.

**Comment by bhack** [comment]
> This will allow us to preserve the numerical_tests/ directory.

Ok, so what do you mean with this claim? It was meant that there was an hypotesis/evaluation to not preserve numerical_tests?

**Comment by LukeWood** [comment]
> > This will allow us to preserve the numerical_tests/ directory.
> 
> Ok, so what do you mean with this claim? It was meant that there was an hypotesis/evaluation to not preserve numerical_tests?

cocoeval.py uses an averaging of points to compute the area under curve.  I am proposing we use a more sophisticated method for higher mathematical accuracy.  This will produce results different from cocoeval.py because they do not do this.

**Comment by bhack** [comment]
Ok so my reproducibility comment was still valid and related to that comment as I was just forcing that it is better to preserve numerical test compatibility (eventually with an optional flag).

**Comment by LukeWood** [comment]
API wise:

This should mirror the pattern used in the `tf.keras.metrics.AUC` metric.

Users can specify which summation method to use in the AUC interpolation:
https://github.com/keras-team/keras/blob/master/keras/metrics/metrics.py#L1646

**Comment by LukeWood** [comment]
The KerasCV team is attempting to narrow the package scope in order to more efficiently use our limited resources. As such, we are only accepting contributions directly related to items on the 6 month roadmap for the package.

If you feel this feature could be particularly useful, or would like to contribute it - please feel free to re-open this issue and ping LukeWood, ianstenbit or tanzhenyu

Feel free to read more on the roadmap:
https://github.com/keras-team/keras-cv/blob/master/.github/ROADMAP.md
